### PR TITLE
Add cluster autoscaling to cloudstack

### PR DESCRIPTION
**Pull request by jonathanmeier5**
*Issue #, if available:*
Epic: https://github.com/aws/eks-anywhere/issues/3186
Ticket: https://github.com/aws/eks-anywhere/issues/3497

*Description of changes:*
Configures autoscaling for cloudstack

*Testing (if applicable):*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/providers/cloudstack/cloudstack_test.go [hunk 1]
```diff
@@ -247,6 +247,37 @@ func TestProviderGenerateCAPISpecForCreate(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("failed to setup and validate: %v", err)
+	}
+
+	cp, md, err := provider.GenerateCAPISpecForCreate(context.Background(), cluster, clusterSpec)
+	if err != nil {
+		t.Fatalf("failed to generate cluster api spec contents: %v", err)
+	}
+	test.AssertContentToFile(t, string(cp), "testdata/expected_results_main_cp.yaml")
+	test.AssertContentToFile(t, string(md), "testdata/expected_results_main_autoscaling_md.yaml")
```
Comment (maxdrib) [inline pkg/providers/cloudstack/cloudstack_test.go:278]: I think this file hasn't been added to the branch

File: pkg/providers/cloudstack/validator_test.go [hunk 2]
```diff
@@ -16,9 +16,10 @@ import (
 )
 
 const (
-	testClusterConfigMainFilename        = "cluster_main.yaml"
-	testClusterConfigMainWithAZsFilename = "cluster_main_with_availability_zones.yaml"
-	testDataDir                          = "testdata"
+	testClusterConfigMainFilename            = "cluster_main.yaml"
+	testClusterConfigMainAutoscalingFilename = "cluster_main_autoscaling.yaml"
```
Comment (maxdrib) [inline pkg/providers/cloudstack/validator_test.go:20]: I think this file is missing. Also, is it necessary to add a new file? Can't you just take the existing one to be parsed and modify it in the test after it's parsed?

**Comment by maxdrib** [comment]
Has this change been tested end to end at all? Is there any documentation for the autoscaling feature?

File: pkg/providers/cloudstack/cloudstack_test.go [hunk 3]
```diff
@@ -247,6 +247,45 @@ func TestProviderGenerateCAPISpecForCreate(t *testing.T) {
...
+	kubectl := mocks.NewMockProviderKubectlClient(mockCtrl)
+	cluster := &types.Cluster{
+		Name: "test",
+	}
+	clusterSpec := givenClusterSpec(t, testClusterConfigMainFilename)
+
+	datacenterConfig := givenDatacenterConfig(t, testClusterConfigMainFilename)
+	machineConfigs := givenMachineConfigs(t, testClusterConfigMainFilename)
+
+	fmt.Println(len(clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations))
```
Comment (maxdrib) [inline pkg/providers/cloudstack/cloudstack_test.go:263]: Let's avoid using fmt.Println. If you're looking to debug a problem, you can use the GoLand or VSCode debugger instead to look at values in the stack

File: pkg/providers/cloudstack/config/template-md.yaml [hunk 4]
```diff
@@ -115,6 +115,11 @@ metadata:
     cluster.x-k8s.io/cluster-name: {{.clusterName}}
   name: {{.workerNodeGroupName}}
   namespace: {{.eksaSystemNamespace}}
+  {{- if .autoscalingConfig }}
+  annotations:
+    cluster.x-k8s.io/cluster-api-autoscaler-node-group-min-size: "{{ .autoscalingConfig.MinCount }}"
+    cluster.x-k8s.io/cluster-api-autoscaler-node-group-max-size: "{{ .autoscalingConfig.MaxCount }}"
```
Comment (maxdrib) [inline pkg/providers/cloudstack/config/template-md.yaml:121]: Has this been tested end to end with an actual CloudStack cluster? Do we need an e2e test added to verify this functionality?

File: pkg/providers/cloudstack/cloudstack_test.go (hunk 3 above)
Comment (jonathanmeier5) [reply pkg/providers/cloudstack/cloudstack_test.go:263]: Yup, missed this. Cleaning up.

**Review by jiayiwang7 — APPROVED** [review]

**Pull request merged**